prusti-specs/src/specifications/preparser.rs

```diff
@@ -907,46 +953,64 @@ mod tests {
-            assert_error!(parse_ghost_constraint(quote!{ [requires(false)] }), err_invalid_arguments);
-            assert_error!(parse_ghost_constraint(quote!{ }), err_invalid_arguments);
-            assert_error!(parse_ghost_constraint(quote!{T: A }), err_invalid_arguments);
-            assert_error!(parse_ghost_constraint(quote!{T: A, [requires(false)], "nope" }), "expected nested specification in brackets");
-            assert_error!(parse_ghost_constraint(quote!{[requires(false)], T: A }), "expected nested specification in brackets");
+            let err_invalid_bounds = "expected one of: `for`, parentheses, `fn`, `unsafe`, `extern`, identifier, `::`, `<`, square brackets, `*`, `&`, `!`, `impl`, `_`, lifetime";
+            assert_error!(parse_ghost_constraint(quote!{ [requires(false)] }), err_invalid_bounds);
+            assert_error!(parse_ghost_constraint(quote!{ }), format!("unexpected end of input, {}", err_invalid_bounds));
+            assert_error!(parse_ghost_constraint(quote!{T: A }), "expected `,`");
+            assert_error!(parse_ghost_constraint(quote!{T: A, [requires(false)], "nope" }), "unexpected extra tokens");
+            assert_error!(parse_ghost_constraint(quote!{[requires(false)], T: A }), err_invalid_bounds);
             assert_error!(parse_ghost_constraint(quote!{T: A,  }), "expected nested specification in brackets");
             assert_error!(parse_ghost_constraint(quote!{T: A, {} }), "expected nested specification in brackets");
         }
 
         #[test]
         fn multiple_bounds_multiple_specs() {
-            let (bounds, nested_specs) = parse_ghost_constraint(quote!{ T: A+B+Foo<i32>, [requires(true), ensures(false)]}).unwrap();
+            let constraint = parse_ghost_constraint(quote!{ T: A+B+Foo<i32>, [requires(true), ensures(false), pure]}).unwrap();
 
-            assert_eq!(bounds.to_string(), "T : A + B + Foo < i32 >");
-            match &nested_specs[0] {
+            assert_bounds_eq(constraint.trait_bounds, quote!{ T : A + B + Foo < i32 > });
+            match &constraint.specs[0] {
                 NestedSpec::Requires(ts) => assert_eq!(ts.to_string(), "true"),
                 _ => panic!(),
             }
-            match &nested_specs[1] {
+            match &constraint.specs[1] {
                 NestedSpec::Ensures(ts) => assert_eq!(ts.to_string(), "false"),
                 _ => panic!(),
             }
+            assert_matches!(&constraint.specs[2], NestedSpec::Pure);
+            assert_eq!(constraint.specs.len(), 3);
         }
 
         #[test]
         fn no_specs() {
-            let (bounds, nested_specs) = parse_ghost_constraint(quote!{ T: A, []}).unwrap();
-            assert_eq!(bounds.to_string(), "T : A");
-            assert!(nested_specs.is_empty());
+            // TODO: does it even make sense to compare parsed quotes like this? feels a bit like saying 2 = 2
```

Comment:
This test is very simple, so the comparison of the trait bound isn't terribly interesting. For the others it makes more sense, and this way it is consistent.